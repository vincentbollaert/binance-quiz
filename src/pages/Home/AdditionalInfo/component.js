import React from 'react'
import { bool, shape, number, string } from 'prop-types'
import styled from 'styled-components'
import { media, UNIT_LG, FONT_SIZE_MD, SUNSET_ORANGE, UNIT_SM } from '../../../styles'
import { CN_ADDITIONAL_INFO, STYLE_ADDITIONAL_INFO_BOX_SHADOW } from '../shared'
import { SHAPE_QUIZ_QUESTION } from '../shapePropTypes'

const Wrap = styled.div`
  display: ${props => props.isShow ? 'block' : 'none'};
  position: absolute;
  right: ${UNIT_LG};
  margin-left: auto;
  ${props => props.isShowLink && 'margin-right: -0.4rem'}; // ellipsis added spacing
  padding: ${UNIT_SM};
  padding-right: 0;
  max-width: 28rem;
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
  background-color: inherit;
  box-shadow: ${STYLE_ADDITIONAL_INFO_BOX_SHADOW};
  color: ${props => props.accentColor || 'inherit'};

  ${media.sm`
    right: 3.4rem;
    display: block;
    font-size: ${FONT_SIZE_MD};
  `};
`
const TimeoutText = styled.div`
  color: ${SUNSET_ORANGE};
`
const Link = styled.a`
  color: ${SUNSET_ORANGE};

  &:hover {
    text-decoration: underline;
  };
`

const AdditionalInfo = ({
  isTimeout,
  isCorrect,
  isQuizComplete,
  accentColor,
  answer,
  correctAnswer,
  questionMatchingAnswer: { slug, excerpt },
  dummyRandomNumber,
}) => {
  const GLOSSARY_URL = `https://www.binance.vision/glossary/${slug}`
  const isShowLink = !isTimeout && isQuizComplete && answer !== correctAnswer && !isCorrect

  return (
    <Wrap
      isShow={!isQuizComplete || isTimeout}
      isShowLink={isShowLink}
      accentColor={accentColor}
      data-testid="component-additional-info"
      className={CN_ADDITIONAL_INFO}
    >
      {isTimeout && <TimeoutText>No time remaining</TimeoutText>}
      {accentColor !== undefined && (
        <span>
          {isShowLink && <Link href={GLOSSARY_URL} target="_blank">{excerpt}</Link>}
          {!isTimeout && isQuizComplete && answer === correctAnswer && `${dummyRandomNumber}% of users got this right`}
          {!isQuizComplete && `${dummyRandomNumber}% chose this option`}
        </span>
      )}
    </Wrap>
  )
}

AdditionalInfo.propTypes = {
  isTimeout: bool,
  isCorrect: bool.isRequired,
  isQuizComplete: bool.isRequired,
  accentColor: string,
  answer: string.isRequired,
  correctAnswer: string.isRequired,
  questionMatchingAnswer: shape(SHAPE_QUIZ_QUESTION).isRequired,
  dummyRandomNumber: number.isRequired,
}

AdditionalInfo.defaultProps = {
  isTimeout: false,
  accentColor: undefined,
}

export default AdditionalInfo
