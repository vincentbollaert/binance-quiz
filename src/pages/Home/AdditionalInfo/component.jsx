import React from 'react'
import { bool, shape, number } from 'prop-types'
import styled from 'styled-components'
import { media, FONT_SIZE_MD, SELECTIVE_YELLOW, SUNSET_ORANGE, UNIT_LG, UNIT_SM, JET } from '../../../styles'

const Wrap = styled.div`
  display: none;
  position: absolute;
  right: ${UNIT_LG};
  right: 3.4rem;
  margin-left: auto;
  padding: ${UNIT_SM};
  padding-right: 0;
  max-width: 28rem;
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
  background-color: ${JET};
  box-shadow: -22px 0px 17px -4px ${JET};
  color: ${props => props.accentColor || 'inherit'};

  ${media.xsm`
    display: block;
    font-size: ${FONT_SIZE_MD};
  `};
`
const TimeoutText = styled.div`
  color: ${SUNSET_ORANGE};
  font-weight: bold;
`
const Link = styled.a`
  color: inherit;

  &:hover {
    text-decoration: underline;
  };
`

const AdditionalInfo = ({
  isTimeout,
  isCorrect,
  isQuizComplete,
  isSelectedAnswerOption,
  questionMatchingAnswer,
  dummyRandomNumber,
}) => {
  const GLOSSARY_URL = `https://www.binance.vision/glossary/${questionMatchingAnswer.slug}`
  const isSelectedAnswerQuizIncomplete = isSelectedAnswerOption && !isQuizComplete

  return (
    <Wrap accentColor={isSelectedAnswerQuizIncomplete && SELECTIVE_YELLOW} isTimeout={isTimeout}>
      {isTimeout && <TimeoutText>No time remaining</TimeoutText>}
      {isQuizComplete && isSelectedAnswerOption && (
        isCorrect
          ? `${dummyRandomNumber}% of users got this right`
          : <Link href={GLOSSARY_URL} target="_blank">{questionMatchingAnswer.excerpt}</Link>
      )}
      {isSelectedAnswerQuizIncomplete && `${dummyRandomNumber}% of users chose this option`}
    </Wrap>
  )
}

AdditionalInfo.propTypes = {
  isTimeout: bool,
  isCorrect: bool.isRequired,
  isQuizComplete: bool.isRequired,
  isSelectedAnswerOption: bool.isRequired,
  questionMatchingAnswer: shape({}).isRequired,
  dummyRandomNumber: number.isRequired,
}

AdditionalInfo.defaultProps = {
  isTimeout: false,
}

export default AdditionalInfo
