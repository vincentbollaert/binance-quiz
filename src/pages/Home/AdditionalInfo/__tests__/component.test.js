import '@testing-library/jest-dom'
import React from 'react'
import { render } from '@testing-library/react'
import AdditionalInfo from '../component'

describe('AdditionalInfo', () => {
  const props = {
    isTimeout: false,
    isCorrect: false,
    isQuizComplete: false,
    accentColor: undefined,
    answer: 'Maker',
    correctAnswer: 'Maker',
    questionMatchingAnswer: { slug: 'slug', excerpt: 'excerpt' },
    dummyRandomNumber: 22,
  }
  test('matches snapshot', () => {
    const { container } = render(<AdditionalInfo {...props} />)
    expect(container.firstChild).toMatchSnapshot()
  })
  test('renders elements conditionally', () => {
    const { container, queryByTestId, queryByText, rerender } = render(<AdditionalInfo {...props} />)
    expect(queryByTestId('component-additional-info')).toBeTruthy()
    expect(queryByText('22% chose this option')).toBeFalsy()

    rerender(<AdditionalInfo {...props} isQuizComplete accentColor="red" />)
    expect(queryByText('22% of users got this right')).toBeTruthy()

    rerender(<AdditionalInfo {...props} accentColor="red" />)
    expect(queryByText('22% chose this option')).toBeTruthy()

    rerender(<AdditionalInfo {...props} accentColor="red" isQuizComplete answer="something" />)
    expect(queryByText('excerpt')).toBeTruthy()
    expect(container.querySelector('a')).toHaveAttribute('href', 'https://www.binance.vision/glossary/slug')
  })
})
