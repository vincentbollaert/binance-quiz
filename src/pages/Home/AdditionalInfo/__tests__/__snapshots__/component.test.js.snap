// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AdditionalInfo matches snapshot 1`] = `
<div
  class="sc-AxjAm bGCVHy additional-info"
  data-testid="component-additional-info"
/>
`;
